crates/sui-framework/sources/governance/voting_power.move

```diff
@@ -0,0 +1,182 @@
+        let i = 0;
+        let len = vector::length(validators);
+        let total_power = 0;
+        let result = vector[];
+        while (i < len) {
+            let validator = vector::borrow(validators, i);
+            let stake = validator::total_stake(validator);
+            let adjusted_stake = (stake as u128) * (TOTAL_VOTING_POWER as u128) / (total_stake as u128);
+            let voting_power = math::min((adjusted_stake as u64), threshold);
+            let info = VotingPowerInfo {
+                validator_index: i,
+                voting_power,
+            };
+            vector::push_back(&mut result, info);
+            total_power = total_power + voting_power;
+            i = i + 1;
+        };
+        (result, TOTAL_VOTING_POWER - total_power)
+    }
+
+    /// Sort the voting power info list, using the voting power, in descending order.
+    fun bubble_sort(info_list: &mut vector<VotingPowerInfo>) {
+        let len = vector::length(info_list);
+        let changed = true;
+        while (changed) {
+            changed = false;
+            let i = 0;
+            while (i + 1 < len) {
+                if (vector::borrow(info_list, i).voting_power < vector::borrow(info_list, i + 1).voting_power) {
+                    changed = true;
+                    vector::swap(info_list, i, i + 1);
+                };
+                i = i + 1;
+            }
+        }
+    }
+
+    /// Distribute remaining_power to validators that are not capped at threshold.
+    fun adjust_voting_power(info_list: &mut vector<VotingPowerInfo>, threshold: u64, remaining_power: u64) {
+        let i = 0;
+        let len = vector::length(info_list);
+        while (i < len && remaining_power > 0) {
+            let v = vector::borrow_mut(info_list, i);
+            let planned = divide_and_round_up(remaining_power, len - i);
+            let target = math::min(threshold, v.voting_power + planned);
+            let actural = math::min(remaining_power, target - v.voting_power);
+            v.voting_power = v.voting_power + actural;
+            assert!(v.voting_power <= threshold, EVotingPowerOverThreshold);
+            remaining_power = remaining_power - actural;
+            i = i + 1;
+        };
+        assert!(remaining_power == 0, ETotalPowerMismatch);
+    }
+
+    /// Update validators with the decided voting power.
+    fun update_voting_power(validators: &mut vector<Validator>, info_list: vector<VotingPowerInfo>) {
+        while (!vector::is_empty(&info_list)) {
+            let VotingPowerInfo {
+                validator_index,
+                voting_power,
+            } = vector::pop_back(&mut info_list);
+            let v = vector::borrow_mut(validators, validator_index);
+            validator::set_voting_power(v, voting_power);
+        };
+        vector::destroy_empty(info_list);
+    }
+
+    /// Calculate x / y, but round up the result.
+    fun divide_and_round_up(x: u64, y: u64): u64 {
```

Comment:
This seems like a useful addition to `math` (I've definitely seen this reimplemented elsewhere).